state-chain/pallets/cf-swapping/src/lib.rs

```diff
@@ -243,27 +247,93 @@ pub struct CcmSwapAmounts {
 }
 
 #[derive(Clone, Debug, PartialEq, Eq, Encode, Decode, TypeInfo)]
-enum CcmSwapState {
-	PrincipalSwapScheduled {
-		principal_swap_id: SwapId,
-		gas_budget: AssetAmount,
-		other_gas_asset: Option<Asset>,
-	},
-	GasSwapScheduled {
-		gas_swap_id: SwapId,
-		principal_output_amount: AssetAmount,
-	},
+enum GasSwapState {
+	OutputReady { gas_budget: AssetAmount },
+	Scheduled { gas_swap_id: SwapId },
+	ToBeScheduled { gas_budget: AssetAmount, other_gas_asset: Asset },
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Encode, Decode, TypeInfo)]
+struct DcaState {
+	scheduled_chunk_swap_id: Option<SwapId>,
+	remaining_input_amount: AssetAmount,
+	remaining_chunks: u32,
+	chunk_interval: u32,
+	accumulated_output_amount: AssetAmount,
+}
+
+impl DcaState {
+	// Create initial DCA state and prepares the first chunk for scheduling; if no dca parameters
+	// provided (for non-DCA swaps), this creates state equivalent to 1 chunk DCA
+	fn create_with_first_chunk(
+		input_amount: AssetAmount,
+		params: Option<DcaParameters>,
+	) -> (DcaState, AssetAmount) {
+		let mut state = DcaState {
+			scheduled_chunk_swap_id: None,
+			remaining_input_amount: input_amount,
+			remaining_chunks: params.as_ref().map(|p| p.number_of_chunks).unwrap_or(1),
+			// Chunk interval won't be used for non-DCA swaps but seems nicer to
+			// set a reasonable default than unwrap Option when it is needed:
+			chunk_interval: params.as_ref().map(|p| p.chunk_interval).unwrap_or(SWAP_DELAY_BLOCKS),
+			accumulated_output_amount: 0,
+		};
+
+		let first_chunk_amount = state.prepare_next_chunk(None).unwrap_or_else(|| {
+			log_or_panic!("Invariant violation: initial DCA state must have at least one chunk!");
+			0
+		});
+
+		(state, first_chunk_amount)
+	}
+
+	fn prepare_next_chunk(
+		&mut self,
+		prev_chunk_and_output: Option<(SwapId, AssetAmount)>,
+	) -> Option<AssetAmount> {
+		if let Some((prev_chunk_swap_id, prev_chunk_output_amount)) = prev_chunk_and_output {
+			if let Some(scheduled_swap_id) = self.scheduled_chunk_swap_id {
```

Comment:
Would it make sense to remove from the scheduled_chunk_swap_id? (Since it's no longer scheduled). 

```suggestion
			if let Some(scheduled_swap_id) = self.scheduled_chunk_swap_id.take() {
```

Reply:
Yes! Updated.